pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+			return err
+		},
+		getClusterWorkspace: func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return kcpClusterClient.Cluster(workspace).TenancyV1alpha1().ClusterWorkspaces().Get(ctx, name, metav1.GetOptions{})
+		},
+	}
+}
+
+type localInformersAccess struct {
+	getClusterWorkspace      func(logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	searchClusterRole        func(lcluster logicalcluster.Name, name string) (bool, error)
+	searchClusterRoleBinding func(lcluster logicalcluster.Name, name string) (bool, error)
+	synced                   func() bool
+}
+
+func buildLocalInformersAccess(kubeSharedInformerFactory coreexternalversions.SharedInformerFactory, kcpSharedInformerFactory kcpexternalversions.SharedInformerFactory) localInformersAccess {
+	clusterWorkspaceInformer := kcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaces().Informer()
+	crInformer := kubeSharedInformerFactory.Rbac().V1().ClusterRoles().Informer()
+	crbInformer := kubeSharedInformerFactory.Rbac().V1().ClusterRoleBindings().Informer()
+	clusterWorkspaceLister := kcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaces().Lister()
+	crLister := kubeSharedInformerFactory.Rbac().V1().ClusterRoles().Lister()
+	crbLister := kubeSharedInformerFactory.Rbac().V1().ClusterRoleBindings().Lister()
+
+	return localInformersAccess{
+		getClusterWorkspace: func(logicalCluster logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			parentLogicalCluster, workspaceName := logicalCluster.Split()
+			result, err := clusterWorkspaceLister.Get(clusters.ToClusterAwareKey(parentLogicalCluster, workspaceName))
+			if err != nil {
+				return nil, err
+			}
+			return result, nil
+		},
+		searchClusterRole: func(workspace logicalcluster.Name, name string) (bool, error) {
+			_, err := crLister.Get(clusters.ToClusterAwareKey(workspace, name))
+			if kerrors.IsNotFound(err) {
+				return false, nil
+			}
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		},
+		searchClusterRoleBinding: func(workspace logicalcluster.Name, name string) (bool, error) {
+			_, err := crbLister.Get(clusters.ToClusterAwareKey(workspace, name))
+			if kerrors.IsNotFound(err) {
+				return false, nil
+			}
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		},
+		synced: func() bool {
+			return clusterWorkspaceInformer.HasSynced() &&
+				crInformer.HasSynced() &&
+				crbInformer.HasSynced()
+		},
+	}
+}
+
+type homeWorkspaceHandlerBuilder struct {
+	apiHandler http.Handler
+
+	externalHost             string
+	homePrefix               logicalcluster.Name
+	bucketLevels, bucketSize int
+	creationDelaySeconds     int
+
+	authz authorizer.Authorizer
+
+	kcp            externalKubeClientsAccess
+	localInformers localInformersAccess
+}
+
+type homeWorkspaceFeatureLogic struct {
```

Comment:
Why is this a struct instead of an interface? What is the utility of injecting each of these separately? What are the cases where we inject some but not all? Are there dependencies between choices for injection here?

Reply:
> Why is this a struct instead of an interface?

This struct contains the functions we want to be possibly overridden in unit tests, in order to cut dependency between functions, and have 1 test == one function.

> What is the utility of injecting each of these separately?

They are all injected as a whole in the `homeWorkspaceHandlerBuilder.build()` method.

> What are the cases where we inject some but not all?

In tests there are cases where we want to override one but not the others

More generally, this allows easier unit testing of individual methods, and is already the result of answering @sttts PR comment after a first try.
IMHO it allows combining the the easiness of testing with the readability of the main handler logic.

However if you have a better approach, contributions are welcomed ... in a followup PR if possible ;-)

Reply:
all this is implementation detail of this package, not an interface to the outside. So I think either way is fine. If this was an API a interface would be natural.